packages/config/src/class-textdomain-customizer.php

```diff
@@ -0,0 +1,213 @@
+	}
+
+	/**
+	 * Find a package in the local composer repo with the input package name. No
+	 * verison constraints are used to find the package.
+	 *
+	 * @param string $package_name The package name.
+	 *
+	 * @return Composer\Package|null The package object.
+	 */
+	private function find_package( $package_name ) {
+		$local_repo = $this->composer->getRepositoryManager()->getLocalRepository();
+		return $local_repo->findPackage( $package_name, '*' );
+	}
+
+	/**
+	 * Traverses the project's packages. Sets the textdomain in the Jetpack
+	 * packages that declare translatable files in their composer.json files.
+	 */
+	public function customize_textdomain_in_packages() {
+		$packages = $this->get_packages();
+
+		if ( ! is_array( $packages ) ) {
+			return;
+		}
+
+		$this->set_vendor_dir();
+		$this->set_textdomain();
+
+		foreach ( $packages as $package ) {
+			$jetpack_package_prefix = 'automattic/jetpack-';
+			$current_package_prefix = substr( $package->getName(), 0, strlen( $jetpack_package_prefix ) );
+
+			if ( $jetpack_package_prefix !== $current_package_prefix ) {
+				// Not a Jetpack package, so skip.
+				continue;
+			}
+
+			if ( $package->getExtra() && isset( $package->getExtra()['translatable'] ) ) {
+				$files = (array) $package->getExtra()['translatable'];
+				$this->customize_textdomain_in_files( $files );
+			}
+		}
+	}
+
+	/**
+	 * Gets the packages in the local composer repository.
+	 *
+	 * @return array of Composer\Package\PackagesInterface objects.
+	 */
+	protected function get_packages() {
+		$local_repo = $this->composer->getRepositoryManager()->getLocalRepository();
+		return $local_repo->getPackages();
+	}
+
+	/**
+	 * Sets the $vendor_dir instance variable using the vendor directory from
+	 * the composer config.
+	 */
+	protected function set_vendor_dir() {
+		$this->vendor_dir = $this->composer->getConfig()->get( 'vendor-dir' );
+	}
+
+	/**
+	 * Sets the textdomain instance variable using the textdomain provided by
+	 * the plugin. If the plugin did not provide a textdomain, uses 'default'.
+	 */
+	private function set_textdomain() {
+		$root_extra = $this->get_root_extra();
+
+		if ( isset( $root_extra['textdomain'] ) &&
+			is_string( $root_extra['textdomain'] ) ) {
+
+				$this->textdomain = $root_extra['textdomain'];
+				return;
+		}
+
+		$this->textdomain = 'default';
+	}
+
+	/**
+	 * Returns the value of the extra section in the root packages's
+	 * composer.json file.
+	 *
+	 * @return array mixed The root package's extra array.
+	 */
+	protected function get_root_extra() {
+		return $this->composer->getPackage()->getExtra();
+	}
+
+	/**
+	 * Sets the textdomain in the input files and directories.
+	 *
+	 * @param array $files The array of translatable files.
+	 */
+	private function customize_textdomain_in_files( $files ) {
+		if ( ! is_array( $files ) ) {
+			return;
+		}
+
+		foreach ( $files as $file ) {
+			$file_path = $this->vendor_dir . $file;
+
+			if ( ! file_exists( $file_path ) ) {
```

Comment:
We have three different places here that check for either `file_exists` or `is_file`, maybe that all can be moved to `customize_textdomain_in_file`? This will also help with the "cognitive complexity" in CodeClimate analysis.